File: src/com/nadmm/airports/wx/MetarParser.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.wx;

import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;

import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.Attributes;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.XMLReader;
import org.xml.sax.helpers.DefaultHandler;

import android.text.format.Time;
import android.util.TimeFormatException;

import com.nadmm.airports.wx.Metar.Flags;

public final class MetarParser {

    public void parse( File xml, Metar metar ) {
        try {
            metar.fetchTime = xml.lastModified();
            SAXParserFactory factory = SAXParserFactory.newInstance();
            SAXParser parser = factory.newSAXParser();
            XMLReader xmlReader = parser.getXMLReader();
            MetarHandler handler = new MetarHandler( metar );
            xmlReader.setContentHandler( handler );
            InputSource input = new InputSource( new FileReader( xml ) );

            xmlReader.parse( input );
        } catch ( Exception e ) {
        }
    }

    protected void parseWxGroups( Metar metar, String wxString ) {
        String[] groups = wxString.split( "\\s+" );
        ArrayList<String> names = WxSymbol.getNames();
        for ( String group : groups ) {
            int offset = 0;
            String intensity = "";
            if ( group.charAt( offset ) == '+' || group.charAt( offset ) == '-' ) {
                intensity = group.substring( offset, offset+1 );
                ++offset;
            }
            while ( offset < group.length() ) {
                WxSymbol wx = null;
                for  ( String name : names  ) {
                    if ( group.substring( offset ).startsWith( name ) ) {
                        wx = WxSymbol.get( name );
                        if ( intensity.length() > 0 ) {
                            wx.setIntensity( intensity );
                            intensity = "";
                        }
                        metar.wxList.add( wx );
                        offset += wx.getName().length();
                        break;
                    }
                }

                if ( wx == null ) {
                    // No match found, skip to next character and try again
                    ++offset;
                }
            }
        }
    }

    protected void parseRemarks( Metar metar ) {
        int index = metar.rawText.indexOf( "RMK" );
        if ( index == -1 ) {
            return;
        }

        String[] rmks = metar.rawText.substring( index ).split( "\\s+" );
        index = 0;
        while ( index < rmks.length ) {
            String rmk = rmks[ index++ ];
            if ( rmk.equals( "PRESRR" ) ) {
                metar.presrr = true;
            } else if ( rmk.equals( "PRESFR" ) ) {
                metar.presfr = true;
            } else if ( rmk.equals( "SNINCR" ) ) {
                metar.snincr = true;
            } else if ( rmk.equals( "WSHFT" ) ) {
                metar.wshft = true;
            } else if ( rmk.equals( "FROPA" ) ) {
                metar.fropa = true;
            } else if ( rmk.equals( "PNO" ) ) {
                metar.flags.add( Flags.RainSensorOff );
            } else if ( rmk.equals( "PK" ) ) {
                rmk = rmks[ index++ ];
                if ( rmk.equals( "WND" ) ) {
                    rmk = rmks[ index++ ];
                    String speed = rmk.substring( 3, rmk.indexOf( '/' ) );
                    metar.windPeakKnots = Integer.valueOf( speed );
                }
            }
        }
    }

    protected final class MetarHandler extends DefaultHandler {

        protected Metar metar;
        private StringBuilder text = new StringBuilder();

        public MetarHandler( Metar metar ) {
            this.metar = metar;
        }

        @Override
        public void characters( char[] ch, int start, int length )
                throws SAXException {
            text.append( ch, start, length );
        }

        @Override
        public void startElement( String uri, String localName, String qName,
                Attributes attributes ) throws SAXException {
            if ( qName.equalsIgnoreCase( "metar" ) ) {
            } else if ( qName.equalsIgnoreCase( "sky_condition" ) ) {
                String name = attributes.getValue( "sky_cover" );
                int cloudBase = 0;
                if ( attributes.getIndex( "cloud_base_ft_agl" ) >= 0 ) {
                    cloudBase = Integer.valueOf( attributes.getValue( "cloud_base_ft_agl" ) );
                }
                SkyCondition skyCondition = SkyCondition.create( name, cloudBase );
                metar.skyConditions.add( skyCondition );
            } else {
                text.delete( 0, text.length() );
            }
        }

        @Override
        public void endElement( String uri, String localName, String qName )
                throws SAXException {
            if ( qName.equalsIgnoreCase( "raw_text" ) ) {
                metar.rawText = text.toString();
                parseRemarks( metar );
            } else if ( qName.equalsIgnoreCase( "station_id" ) ) {
                metar.stationId = text.toString();
            } else if ( qName.equalsIgnoreCase( "observation_time" ) ) {
                try {
                    Time time = new Time();
                    time.parse3339( text.toString() );
                    metar.observationTime = time.toMillis( true );
                } catch ( TimeFormatException e ) {
                }
            } else if ( qName.equalsIgnoreCase( "elevation_m" ) ) {
                metar.stationElevationMeters = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "temp_c" ) ) {
                metar.tempCelsius = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "dewpoint_c" ) ) {
                metar.dewpointCelsius = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "wind_dir_degrees" ) ) {
                metar.windDirDegrees = Integer.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "wind_speed_kt" ) ) {
                metar.windSpeedKnots = Integer.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "wind_gust_kt" ) ) {
                metar.windGustKnots = Integer.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "visibility_statute_mi" ) ) {
                metar.visibilitySM = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "altim_in_hg" ) ) {
                metar.altimeterHg = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "sea_level_pressure_mb" ) ) {
                metar.seaLevelPressureMb = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "corrected" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.Corrected ) ;
                }
            } else if ( qName.equalsIgnoreCase( "auto_station" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.AutoStation ) ;
                }
            } else if ( qName.equalsIgnoreCase( "auto" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.AutoReport ) ;
                }
            } else if ( qName.equalsIgnoreCase( "maintenance_indicator_on" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.MaintenanceIndicatorOn ) ;
                }
            } else if ( qName.equalsIgnoreCase( "present_weather_sensor_off" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.PresentWeatherSensorOff ) ;
                }
            } else if ( qName.equalsIgnoreCase( "lightning_sensor_off" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.LightningSensorOff ) ;
                }
            } else if ( qName.equalsIgnoreCase( "freezing_rain_sensor_off" ) ) {
                if ( text.toString().equalsIgnoreCase( "true" ) ) {
                    metar.flags.add( Flags.FreezingRainSensorOff ) ;
                }
            } else if ( qName.equalsIgnoreCase( "wx_string" ) ) {
                parseWxGroups( metar, text.toString() );
            } else if ( qName.equalsIgnoreCase( "flight_category" ) ) {
                metar.flightCategory = text.toString();
            } else if ( qName.equalsIgnoreCase( "three_hr_pressure_tendency_mb" ) ) {
                metar.pressureTend3HrMb = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "maxt_c" ) ) {
                metar.maxTemp6HrCentigrade = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "mint_c" ) ) {
                metar.minTemp6HrCentigrade = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "maxt24hr_c" ) ) {
                metar.maxTemp24HrCentigrade = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "mint24hr_c" ) ) {
                metar.minTemp24HrCentigrade = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "precip_in" ) ) {
                metar.precipInches = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "pcp3hr_in" ) ) {
                metar.precip3HrInches = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "pcp6hr_in" ) ) {
                metar.precip6HrInches = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "pcp24hr_in" ) ) {
                metar.precip24HrInches = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "snow_in" ) ) {
                metar.snowInches = Float.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "vert_vis_ft" ) ) {
                metar.vertVisibilityFeet = Integer.valueOf( text.toString() );
            } else if ( qName.equalsIgnoreCase( "metar_type" ) ) {
                metar.metarType = text.toString();
            } else if ( qName.equalsIgnoreCase( "metar" ) ) {
                metar.isValid = true;
                metar.setMissingFields();
            }
        }
    }

}

File: src/main/java/com/nadmm/airports/wx/MetarParser.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011-2023 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import com.nadmm.airports.utils.WxUtils
import com.nadmm.airports.utils.WxUtils.computeFlightCategory
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler
import java.io.File
import java.time.Instant
import java.time.format.DateTimeParseException
import javax.xml.parsers.SAXParserFactory

class MetarParser {

    fun parse(xmlFile: File, stationIds: List<String>): List<Metar> {
        val parsedMetars = mutableMapOf<String, Metar>()

        val handler = MetarHandler(xmlFile, parsedMetars)
        val factory = SAXParserFactory.newInstance()
        factory.newSAXParser().parse(xmlFile, handler)

        // Now put the missing ones
        stationIds.forEach { stationId ->
            parsedMetars.getOrPut(stationId) {
                Metar(stationId = stationId)
            }
        }

        return parsedMetars.values.toList()
    }

    private inner class MetarHandler(
        private val xmlFile: File,
        private val parsedMetars: MutableMap<String, Metar>
    ) : DefaultHandler() {
        private lateinit var metar: Metar
        private val sb = StringBuilder()

        override fun characters(ch: CharArray, start: Int, length: Int) {
            sb.append(ch, start, length)
        }

        override fun startElement(
            uri: String, localName: String, qName: String,
            attributes: Attributes
        ) {
            sb.clear()
            if (qName.equals("metar", ignoreCase = true)) {
                metar = Metar(fetchTime = xmlFile.lastModified())
            } else if (qName.equals("sky_condition", ignoreCase = true)) {
                val skyCover = attributes.getValue("sky_cover")
                val cloudBaseAGL = attributes.getValue("cloud_base_ft_agl")?.toInt() ?: 0
                val skyCondition = SkyCondition.of(skyCover, cloudBaseAGL)
                metar.skyConditions.add(skyCondition)
            }
        }

        // In endElement function
        override fun endElement(uri: String, localName: String, qName: String) {
            val text = sb.trim().toString()

            when (qName.lowercase()) {
                "metar" -> {
                    metar.isValid = metar.stationId != null && metar.rawText != null
                    if (metar.isValid) {
                        parseRemarks(metar)
                        setMissingFields(metar)
                    }
                    metar.stationId?.let { parsedMetars[it] = metar }
                }
                "raw_text" -> metar.rawText = text
                "observation_time" -> {
                    try {
                        val time = Instant.parse(text)
                        metar.observationTime = time.toEpochMilli()
                    } catch (_: DateTimeParseException) { // Use a more specific exception
                    }
                }
                "station_id" -> metar.stationId = text
                "elevation_m" -> metar.stationElevationMeters = text.toFloatOrNull() ?: 0.0f
                "temp_c" -> metar.tempCelsius = text.toFloatOrNull() ?: 0.0f
                "dewpoint_c" -> metar.dewpointCelsius = text.toFloatOrNull() ?: 0.0f
                "wind_dir_degrees" -> metar.windDirDegrees = text.toIntOrNull() ?: 0
                "wind_speed_kt" -> metar.windSpeedKnots = text.toIntOrNull() ?: 0
                "wind_gust_kt" -> metar.windGustKnots = text.toIntOrNull() ?: 0
                "visibility_statute_mi" -> metar.visibilitySM = text.toFloatOrNull() ?: 10.1f
                "altim_in_hg" -> metar.altimeterHg = text.toFloatOrNull() ?: 0.0f
                "sea_level_pressure_mb" -> metar.seaLevelPressureMb = text.toFloatOrNull() ?: 0.0f
                "wx_string" -> WxSymbol.parseWxSymbols(metar.wxList, text)
                "flight_category" -> metar.flightCategory = text
                "three_hr_pressure_tendency_mb" -> metar.pressureTend3HrMb = text.toFloatOrNull() ?: 0.0f
                "maxt_c" -> metar.maxTempCelsiusLast6Hours = text.toFloatOrNull() ?: 0.0f
                "mint_c" -> metar.minTempCelsiusLast6Hours = text.toFloatOrNull() ?: 0.0f
                "maxt24hr_c" -> metar.maxTempCelsiusLast24Hours = text.toFloatOrNull() ?: 0.0f
                "mint24hr_c" -> metar.minTempCelsiusLast24Hours = text.toFloatOrNull() ?: 0.0f
                "precip_in" -> metar.precipInches = text.toFloatOrNull() ?: 0.0f
                "pcp3hr_in" -> metar.precip3HrInches = text.toFloatOrNull() ?: 0.0f
                "pcp6hr_in" -> metar.precip6HrInches = text.toFloatOrNull() ?: 0.0f
                "pcp24hr_in" -> metar.precip24HrInches = text.toFloatOrNull() ?: 0.0f
                "snow_in" -> metar.snowInches = text.toFloatOrNull() ?: 0.0f
                "vert_vis_ft" -> metar.vertVisibilityFeet = text.toIntOrNull() ?: Int.MAX_VALUE
                "metar_type" -> metar.metarType = text

                // Flags
                "auto", "auto_station" -> if (text.isTrue()) metar.flags.add(MetarFlag.AUTOMATED_STATION)
                "corrected" -> if (text.isTrue()) metar.flags.add(MetarFlag.CORRECTED)
                "maintenance_indicator_on" -> if (text.isTrue()) metar.flags.add(MetarFlag.MAINTENANCE_INDICATOR_ON)
                "present_weather_sensor_off" -> if (text.isTrue()) metar.flags.add(MetarFlag.PRESENT_WEATHER_SENSOR_OFF)
                "lightning_sensor_off" -> if (text.isTrue()) metar.flags.add(MetarFlag.LIGHTNING_SENSOR_OFF)
                "freezing_rain_sensor_off" -> if (text.isTrue()) metar.flags.add(MetarFlag.FREEZING_RAIN_SENSOR_OFF)
                "no_signal" -> if (text.isTrue()) metar.flags.add(MetarFlag.NO_SIGNAL)
            }
        }
    }

    private fun parseRemarks(metar: Metar) {
        val rawText = metar.rawText ?: return
        val index = rawText.indexOf("RMK")
        if (index == -1) {
            return
        }
        val remarks = rawText.substring(index+4)
            .split("\\s+".toRegex())
            .dropLastWhile { it.isEmpty() }
        val it = remarks.iterator()
        while (it.hasNext()) {
            var rmk = it.next()
            when (rmk) {
                "PRESRR" -> metar.weatherPhenomena.add(WeatherPhenomenon.PRESSURE_RISING_RAPIDLY)
                "PRESFR" -> metar.weatherPhenomena.add(WeatherPhenomenon.PRESSURE_FALLING_RAPIDLY)
                "SNINCR" -> metar.weatherPhenomena.add(WeatherPhenomenon.SNOW_INCREASING_RAPIDLY)
                "WSHFT" -> metar.weatherPhenomena.add(WeatherPhenomenon.WIND_SHIFT)
                "FROPA" -> metar.weatherPhenomena.add(WeatherPhenomenon.FRONTAL_PASSAGE)
                "LTG" -> metar.weatherPhenomena.add(WeatherPhenomenon.LIGHTNING)
                "PNO" -> metar.flags.add(MetarFlag.RAIN_SENSOR_OFF)
                "PK" -> {
                    rmk = if (it.hasNext()) it.next() else ""
                    if (rmk == "WND" && it.hasNext()) {
                        rmk = it.next()
                        metar.windPeakKnots = rmk.substring(3, rmk.indexOf('/')).toIntOrNull() ?: 0
                    }
                }
            }

        }
    }

    private fun setMissingFields(metar: Metar) {
        if (metar.flightCategory == WxUtils.FLIGHT_CATEGORY_UNKN) {
            metar.flightCategory = computeFlightCategory(metar.skyConditions, metar.visibilitySM)
        }
        if (metar.vertVisibilityFeet < Int.MAX_VALUE) {
            // Check to see if we have an OVX layer, if not add it
            metar.skyConditions.find { sky -> sky.skyCover == "OVX" } ?:
                metar.skyConditions.add(SkyCondition.of("OVX", 0))
        }
        if (metar.skyConditions.isEmpty()) {
            // Sky condition is not available in the METAR
            metar.skyConditions.add(SkyCondition.of("SKM", 0))
        }
    }

    private fun CharSequence.isTrue(): Boolean {
        return this.toString().equals("true", ignoreCase = true)
    }
}